### PR TITLE
[RFC] Use sampler-based Random API

### PR DESCRIPTION
**Pull request by kimikage**
This is a proposal to drastically change the implementation of `rand`. (cf. https://github.com/JuliaGraphics/ColorTypes.jl/issues/150#issuecomment-594986636)

### CODE REVIEW

**Comment by kimikage** [comment]
At this time, I have not applied the `rand!` optimization using `reinterpret` or `unsafe_wrap`, but it is not fatally slow.
```julia
julia> @btime rand(RGB{Float32}, 1000, 1000, 10);
  97.308 ms (2 allocations: 114.44 MiB) # v1.6.0-beta1 Windows
  95.061 ms (2 allocations: 114.44 MiB) # v1.6.0-beta1 Debian on WSL2
  134.923 ms (2 allocations: 114.44 MiB) # v1.5.3 Windows

julia> @btime rand(HSV{Float32}, 1000, 1000, 10);
  124.078 ms (2 allocations: 114.44 MiB) # v1.6.0-beta1 Windows
  122.759 ms (2 allocations: 114.44 MiB) # v1.6.0-beta1 Debian on WSL2
  152.048 ms (2 allocations: 114.44 MiB) # v1.5.3 Windows
```

**Comment by kimikage** [comment]
I'm not sure the reason, but precompilation seems to be of little help.

File: src/operations.jl [hunk 1]
```diff
@@ -13,80 +13,105 @@ gamutmax(::Type{T}) where {T<:HSL} = (360,1,1)
...
+    rand!(r, A, SamplerType{base_colorant_type(C){Float64}}())
+end
+function rand!(r::AbstractRNG, A::Array{C}, ::SamplerType{C}) where {T, C<:Colorant{T}}
+    A .= rand.((r,), C)
+end
+function _rand01!(r::AbstractRNG, A::Array{C},
+               ::SamplerType{C}) where {T<:Rand01Type, C<:Colorant{T}}
+    N = sizeof(C) ÷ sizeof(T)
+    T0 = T <: FixedPoint ? FixedPointNumbers.rawtype(T) : T
+    At = unsafe_wrap(Array, reinterpret(Ptr{T0}, pointer(A)), (N, size(A)...))
```
Comment (timholy) [inline src/operations.jl:79]: Are you using `unsafe_wrap` to bypass the performance problem of `ReinterpretArray` prior to Julia 1.6? I'm fine with that, I just remember someone complaining about that solution previously.

**Review by timholy — APPROVED** [review]
I haven't reviewed this in gory detail, but it looks like you're adding more flexibility by supporting different samplers, which seems like a win to me and a very nice change.

File: src/operations.jl (hunk 1 above)
Comment (kimikage) [reply src/operations.jl:79]: No, it has nothing to do with `ReinterpretArray` here. (Of course, therein lies the root problem.)
It is an issue with using the private API.
https://github.com/JuliaLang/julia/issues/36718

**Comment by kimikage** [comment]
In any case, this is too big a change to include in a patch release.

**Pull request merged**